src/VisualStudio/Core/Def/Options/VisualStudioOptionStorage.cs

```diff
@@ -0,0 +1,422 @@
+
+        /// <summary>
+        /// VB specific key should only be specified for backward compat for a few speciifc options.
+        /// Language specific storage key should use %LANGUAGE% placeholder in <see cref="Key"/>.
+        /// </summary>
+        public string? VisualBasicKey { get; }
+
+        public RoamingProfileStorage(string key)
+        {
+            Key = key;
+        }
+
+        /// <summary>
+        /// Backward compat only.
+        /// </summary>
+        [Obsolete]
+        public RoamingProfileStorage(string key, string vbKey)
+        {
+            Key = key;
+            VisualBasicKey = vbKey;
+        }
+
+        private string GetKey(string? language)
+            => (VisualBasicKey != null && language == LanguageNames.VisualBasic) ? VisualBasicKey : SubstituteLanguage(Key, language);
+
+        private static string SubstituteLanguage(string keyName, string? language)
+            => keyName.Replace("%LANGUAGE%", language switch
+            {
+                LanguageNames.CSharp => "CSharp",
+                LanguageNames.VisualBasic => "VisualBasic",
+                _ => language // handles F#, TypeScript and Xaml
+            });
+
+        public bool TryPersist(VisualStudioSettingsOptionPersister persister, OptionKey2 optionKey, object? value)
+            => persister.TryPersist(optionKey, GetKey(optionKey.Language), value);
+
+        public bool TryFetch(VisualStudioSettingsOptionPersister persister, OptionKey2 optionKey, out object? value)
+            => persister.TryFetch(optionKey, GetKey(optionKey.Language), out value);
+    }
+
+    internal sealed class FeatureFlagStorage : VisualStudioOptionStorage
+    {
+        public string FlagName { get; }
+
+        public FeatureFlagStorage(string flagName)
+        {
+            FlagName = flagName;
+        }
+
+        public bool TryPersist(FeatureFlagPersister persister, object? value)
+            => persister.TryPersist(FlagName, value);
+
+        public bool TryFetch(FeatureFlagPersister persister, OptionKey2 optionKey, out object? value)
+            => persister.TryFetch(optionKey, FlagName, out value);
+    }
+
+    internal sealed class LocalUserProfileStorage : VisualStudioOptionStorage
+    {
+        private readonly string _path;
+        private readonly string _key;
+
+        public LocalUserProfileStorage(string path, string key)
+        {
+            _path = path;
+            _key = key;
+        }
+
+        public bool TryPersist(LocalUserRegistryOptionPersister persister, OptionKey2 optionKey, object? value)
+            => persister.TryPersist(optionKey, _path, _key, value);
+
+        public bool TryFetch(LocalUserRegistryOptionPersister persister, OptionKey2 optionKey, out object? value)
+            => persister.TryFetch(optionKey, _path, _key, out value);
+    }
+
+    public static readonly IReadOnlyDictionary<string, VisualStudioOptionStorage> Storages = new Dictionary<string, VisualStudioOptionStorage>()
```

Comment:
Whats to prevent someone from adding the same key twice but to different storage locations?

Reply:
`VisualStudioOptionStorageTests` test validates this.

Reply:
nevermind, I see the test for this:

https://github.com/tmat/roslyn/blob/d60e692fbee0bdf57315cb83264274043ceacac3/src/VisualStudio/Core/Test.Next/Options/VisualStudioOptionStorageTests.cs#L34